core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -288,7 +288,7 @@ private[spark] class ExecutorAllocationManager(
     }
 
     // Update executor target number only after initializing flag is unset
-    updateAndSyncNumExecutorsTarget(clock.getTimeMillis())
+    updateAndSyncNumExecutorsTarget(clock.nanoTime())
```

Comment:
System.nanoTime() is not monotonic across CPU cores. Are you sure that this value of `clock.nanoTime()` is not compared to `System.nanoTime()` from another thread? For instance, from `onSchedulerBacklogged()`.

Reply:
Are you sure about that?

This is what the javadocs say:

    The same origin is used by all invocations of this method in an instance of a Java virtual
    machine; other virtual machine instances are likely to use a different origin. 

(Comment about "other virtual machines" aside.)

That seems to be corroborated by other sources, too (e.g. https://stackoverflow.com/questions/510462/is-system-nanotime-completely-useless). Seems in a long gone past there might have been issues, but not today.

Reply:
Also, just wanted to point out that it's ok if `nanoTime` is not completely monotonic, as long as it's better than `getTimeMillis()`. After all, this code was working with a non-monotonic clock before.

The only problem is if different threads can have wildly different views of what `nanoTime` returns, so that comparing the values makes no sense. I don't believe that to be the case.

Reply:
`System.nanoTime()` can use `clock_gettime(CLOCK_MONOTONIC, ...)` on Linux: http://hg.openjdk.java.net/jdk8/jdk8/hotspot/file/tip/src/os/linux/vm/os_linux.cpp#l1453
`CLOCK_MONOTONIC` is based on the Time Stamp Counter (https://en.wikipedia.org/wiki/Time_Stamp_Counter). Its synchronization across CPU cores depends on CPU architecture, for example: _"AMD processors up to the K8 core always incremented the time-stamp counter every clock cycle.[6] Thus, power management features were able to change the number of increments per second, and the values could get out of sync between different cores or processors in the same system."_

Reply:
That is interesting, but according to this:

    http://btorpey.github.io/blog/2014/02/18/clock-sources-in-linux/

`CLOCK_MONOTONIC` uses RDTSC (or RDTSCP when available), which solves that problem. The preferred one (RDTSCP) seems to be available in all modern CPUs. So, `CLOCK_MONOTONIC` actually fulfills the requirements of the `nanoTime()` API (that the source of time is consistent).

(The AMD K8 mentioned in your comment, which came out more than 10 years ago, is where RDTSCP was added.)

Reply:
Probably worth to link directly to the following also (which is linked from the article above)

https://stackoverflow.com/questions/10921210/cpu-tsc-fetch-operation-especially-in-multicore-multi-processor-environment

Reply:
this is a good discussion -- its really easy to get confused about the details here, especially as some googling can take you to old info.  Worth including a bit of this on the doc for `SystemClock.nanoTime()`?  Maybe at least just noting that this is dependable across cores as long as its Java7+.

Reply:
+1 for @squito 's comment. Also, could you summarize the above into the PR description a little bit please, @vanzin ?